digdag-core/src/main/java/io/digdag/core/database/BasicDatabaseStoreManager.java

```diff
@@ -274,42 +283,105 @@ public int getIndex()
-            throw Throwables.propagate(innerException);
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T, E1 extends Exception, E2 extends Exception> T transaction(TransactionActionWithExceptions<T, D, E1, E2> action, Class<E1> exClass1, Class<E2> exClass2) throws E1, E2
+    {
+        Handle nested = currentTransaction.get();
+        if (nested != null) {
+            return action.call(nested, nested.attach(daoIface));
         }
-        catch (InterruptedException ex) {
-            throw Throwables.propagate(ex);
+        else {
+            try {
+                return transactionRetryExecutor.runInterruptible(() -> {
+                    try (Handle handle = dbi.open()) {
+                        currentTransaction.set(handle);
+                        handle.begin();
+                        T retval;
+                        try {
+                            retval = action.call(handle, handle.attach(daoIface));
+                        }
+                        catch (Exception ex) {
+                            try {
+                                handle.rollback();
+                            }
+                            catch (Exception rollback) {
+                                ex.addSuppressed(rollback);
+                            }
+                            Throwables.propagateIfInstanceOf(ex, exClass1);
+                            Throwables.propagateIfInstanceOf(ex, exClass2);
+                            Throwables.propagateIfPossible(ex);
+                            throw new TransactionFailedException(
+                                    "Transaction failed due to exception being thrown " +
+                                    "from within the callback. See cause " +
+                                    "for the original exception.", ex);
+                        }
+                        handle.commit();
+                        return retval;
+                    }
+                    finally {
+                        currentTransaction.set(null);
+                    }
+                });
+            }
+            catch (RetryGiveupException ex) {
+                Throwable innerException = ex.getCause();
+                Throwables.propagateIfInstanceOf(innerException, exClass1);
+                Throwables.propagateIfInstanceOf(innerException, exClass2);
+                throw Throwables.propagate(innerException);
+            }
+            catch (InterruptedException ex) {
+                throw Throwables.propagate(ex);
+            }
         }
     }
 
     @SuppressWarnings("unchecked")
-    public <T, E1 extends Exception, E2 extends Exception> T transaction(TransactionActionWithExceptions<T, D, E1, E2> action, Class<E1> exClass1, Class<E2> exClass2) throws E1, E2
+    public <T, E1 extends Exception, E2 extends Exception> T transactionWithLocalRetry(RetryableTransactionActionWithExceptions<T, D, E1, E2> action, Class<E1> exClass1, Class<E2> exClass2) throws E1, E2
```

Comment:
Purpose of transactionWithLocalRetry is specifically put-and-lock implementations: putAndLockProject and putAndLockSession. These 2 operations want to insert a record if not exist, lock it within the same transaction, and rollback everything if something failed in the transaction.

putAndLockProject is not a big issue because it won't be called within another transaction.

putAndLockSession needs above behavior because session is created only when one doesn't exist. If exists, an attempt must exist. The current implementation for PostgreSQL (<= 9.4) and H2 is to INSERT first and retry the transaction if conflict happens. Because the conflict is an expected situation, here shouldn't raise exceptions.